### DISCUSSION TITLE
Livewire 3: Solving 404 error on livewire/update without access to server configuration

### DISCUSSION THREAD

**Discussion by onlyasmalllizard** [livewire/livewire#7915] [1. Help] [answered]
Hello, sorry for needing to ask for so much help in such a short period of time!

I'm trying to host my Livewire project and I've run into the issue where POST requests to `livewire/update` return a 404. I've been going through previous threads trying to fix the issue with no luck. I was hoping someone here might have some more ideas of things to try.

My current hosting plan is on a shared server, so I can't change any server configurations. It's an Apache 2.4.58 server running on x86_64 Linux. Annoyingly, I also don't have ssh access. The hosting has been provided by an organisation, so I can't immediately switch providers or upgrade the plan. If you think the hosting situation is going to make things needlessly difficult, please let me know!

In order to fix 404s caused by loading livewire.js, I created a `vendor/livewire` folder in the public folder as described in [this comment](https://github.com/livewire/livewire/discussions/6328#discussioncomment-7059979):
```
public
- vendor
- - livewire
- - - livewire.esm.js
- - - livewire.js
- - - manifest.json
```

(In a previous version, I had followed the docs instructions to [manually bundle assets](https://livewire.laravel.com/docs/installation#manually-bundling-livewire-and-alpine), but I was running into [this error](https://github.com/livewire/livewire/discussions/6328) until I switched to the "assets in the public folder" approach. I don't know if this is notable or not, but once I changed to the current approach, the styling of the 404 error modal changed drastically. Neither looked unstyled, just different.)

index.blade.php:
```
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">

        <title>Laravel</title>

        <link rel="preconnect" href="https://fonts.googleapis.com">
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;600&display=swap" rel="stylesheet">

        @vite([
            'resources/styles/main.scss',
            'resources/js/app.js'
        ])

        <script
            defer
            src="https://cdn.jsdelivr.net/npm/a11y-dialog@8/dist/a11y-dialog.min.js"
        ></script>
    </head>
    <body>
        <main>
            <livewire:epdm-calculator />
        </main>
    </body>
</html>
```

app.js:
```
import './bootstrap';
import 'alpinejs';
```

My routes in web.php:
```
Route::get('/', function () {
    return view('index');
});

Livewire::setScriptRoute(function ($handle) {
    return Route::get('/public/vendor/livewire/livewire.js', $handle);
});

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/public/vendor/livewire/update', $handle);
});
```

.htaccess:
```
<IfModule mod_rewrite.c>
    <IfModule mod_negotiation.c>
        Options -MultiViews -Indexes
    </IfModule>

    RewriteEngine On

    # Handle Authorization Header
    RewriteCond %{HTTP:Authorization} .
    RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]

    # Redirect Trailing Slashes If Not A Folder...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_URI} (.+)/$
    RewriteRule ^ %1 [L,R=301]

    # Send Requests To Front Controller...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^ index.php [L]
</IfModule>

<Files .env>
    order allow,deny
    Deny from all
</Files>
```

If you'd like to see the errors in action, the site is [here](https://dfb70396c925852b997745d455ab3a93-18329.sites.k-hosting.co.uk/public/). Interacting with any of the form elements will trigger it.

I'd really appreciate any advice!

**Comment by shubdragon** [comment]
did you check the console ? have you set app url correctly or deleted hot file from your public folder ? but the update-uri not looks okay according to .htaccess config you update uri should be `/public/public/vendor/livewire/update` and it will work (inspect your body tag for livewire script)
![image](https://github.com/livewire/livewire/assets/20332532/e4edc5c5-9a41-486b-9bb8-4fb2e3a848f5)

So your current solution will be this
Solution:1
https://github.com/livewire/livewire/assets/20332532/a42d2845-6de6-476f-8579-76652cbad41b
`
Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/public/public/vendor/livewire/update', $handle);
});
`

Solution: 2 (need to try)
you can modify .htaccess something live this or alternately point domain directly to public folder

`<IfModule mod_rewrite.c>
    <IfModule mod_negotiation.c>
        Options -MultiViews -Indexes
    </IfModule>

    RewriteEngine On

    # Handle Authorization Header
    RewriteCond %{HTTP:Authorization} .
    RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]

    # Redirect Trailing Slashes If Not A Folder...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_URI} (.+)/$
    RewriteRule ^(.*)$ public/$1 
    RewriteRule ^ %1 [L,R=301]

    # Send Requests To Front Controller...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^ index.php [L]
</IfModule>
<Files .env>
       order allow,deny
       Deny from all
</Files>
`

**Reply by onlyasmalllizard** [reply]
Hi @shubdragon, thanks for your help. I did see I had the hot file in there and have deleted it. The app url is set to `https://www.dfb70396c925852b997745d455ab3a93-18329.sites.k-hosting.co.uk`.

When I tried Solution 2, `https://www.dfb70396c925852b997745d455ab3a93-18329.sites.k-hosting.co.uk` showed a 404, and `https://www.dfb70396c925852b997745d455ab3a93-18329.sites.k-hosting.co.uk/public` redirected to  `/public/%251`, which caused a 404 on the main screen. I reverted the .htaccess file to what it originally was.

For Solution 1, I'm a little confused, because I can see it worked for you, but I'm getting the same behaviour as before:
<img width="1464" alt="Screenshot 2024-02-09 at 08 32 13" src="https://github.com/livewire/livewire/assets/30602046/6bb84851-a550-43aa-b4e9-f05b8c0d965e">

I've tried clearing my cache, and I've also tried loading it on my phone, where I've never visited the site before, and the behaviour is the same.

These are my current routes:
```
Route::get('/', function () {
    return view('index');
});

Livewire::setScriptRoute(function ($handle) {
    return Route::get('/public/vendor/livewire/livewire.js', $handle);
});

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/public/public/vendor/livewire/update', $handle);
});
```

**Reply by shubdragon** [reply, marked as answer]
(**Assumed this .htaccess is not which one present in public folder, i assumed this .htaccess file present on your project root with other files e.g .env, if not present just create it or it will be automatically created by shared hosting if you change any config or change php version, sorry i have skipped to mention i am not talking about the .htaccess present in public folder)

just for the instance of testing purpose, 
just comment  the whole  .htaccess config and just use this lines
`
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteRule ^(.*)$ public/$1 [L]
</IfModule>
`
and revert back update url to
`Livewire::setUpdateRoute(function ($handle) { return Route::post('/public/vendor/livewire/update', $handle); });`

Or Probably it should not include public if you modified the .htaccess as above

`Livewire::setScriptRoute(function ($handle) {
    return Route::get(/vendor/livewire/livewire.js', $handle);
});`

`Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/vendor/livewire/update', $handle);
});`


If this work for you, **Or** alternately try pointing domain directly to project public folder if possible

**Reply by onlyasmalllizard** [reply]
I just get a 404 with the .htaccess like that, with any of the suggested route configurations. Is that to be expected, since it's not directing the site to the index.php?

I can't change where this domain points, but I was able to configure a subdomain pointing to the public folder. I'm waiting for it to propagate to test whether that's worked

**Reply by shubdragon** [reply]
lets wait you if pointed subdomain to public folder directly, i forget to mention before i am talking about the .htaccess on root folder, if you do this way you will not need to point subdomain to public folder, did you read above note on last message

**Reply by onlyasmalllizard** [reply]
I think I read it before you added the part about it not being in the public folder. I added a root .htaccess, updated the routes, and it's working now. Thank you so much for your help!

**Reply by shubdragon** [reply]
I have faced this type of problem a lot with shared hosting in the past.. finally it is working... now i am satisfied .. lol 😂😀 : )

now you don't need that /public in your domain url it will work anyway

**Comment by kickymaulana** [comment]
if your url http://localhost/yourproject/public

follow this tutorial https://livewire.laravel.com/docs/installation#manually-including-livewires-frontend-assets

and set 

this code from
Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/custom/livewire/update', $handle);
});

to 

Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/yourproject/public/livewire/update', $handle);
});
in your route web.php

**Comment by corcronje** [comment]
I've just had this issue and found that clearing the routes cache resolved the issue.

`php artisan route:clear`

**Comment by MeiKatz** [comment]
I ran into this problem because I had a property with the type of an Eloquent model that was not nullable and Livewire tried to set it to null while updating. This caused an error in `Livewire\Mechanisms\HandleComponents#hydrateProperties()` and so the Laravel returned a 404 page.

**Reply by pomsclete** [reply]
this response is exact replace eloquent method to query builder this work

**Comment by SyedMuradAliShah** [comment]
I had the same problem, the fix for it was.

-> I just copied `.htaccess` file and paste it into the `public_html` (where the my index.php resides)
